File: DSA/VedantChouhan_590024563/Day-34/q2.c
#include <stdio.h>

#define MAX 1000

int maxVisiblePeople(int arr[], int n) {

    int prev[MAX], next[MAX];
    int stack[MAX], top = -1;

    // Previous Greater or Equal
    for (int i = 0; i < n; i++) {
        while (top != -1 && arr[stack[top]] < arr[i])
            top--;

        if (top == -1)
            prev[i] = -1;
        else
            prev[i] = stack[top];

        stack[++top] = i;
    }

    // Next Greater or Equal
    top = -1;
    for (int i = n - 1; i >= 0; i--) {
        while (top != -1 && arr[stack[top]] < arr[i])
            top--;

        if (top == -1)
            next[i] = n;
        else
            next[i] = stack[top];

        stack[++top] = i;
    }

    int ans = 1;

    for (int i = 0; i < n; i++) {
        int left = i - prev[i] - 1;
        int right = next[i] - i - 1;
        int total = left + right + 1;

        if (total > ans)
            ans = total;
    }

    return ans;
}

int main() {

    int n;

    printf("Enter number of people: ");
    scanf("%d", &n);

    int arr[n];

    printf("Enter heights: ");
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    printf("Maximum people visible: %d\n", maxVisiblePeople(arr, n));

    return 0;
}

File: DSA/Navneet_590012555/Day34/Q2.java
int maxVisible(int[] arr) {
    int n = arr.length;
    int[] left = new int[n];
    int[] right = new int[n];
    Stack<Integer> st = new Stack<>();

    // Visible to the left
    for (int i = 0; i < n; i++) {
        while (!st.isEmpty() && arr[st.peek()] < arr[i])
            st.pop();

        left[i] = st.isEmpty() ? i : i - st.peek();
        st.push(i);
    }

    st.clear();

    // Visible to the right
    for (int i = n - 1; i >= 0; i--) {
        while (!st.isEmpty() && arr[st.peek()] < arr[i])
            st.pop();

        right[i] = st.isEmpty() ? n - 1 - i : st.peek() - i;
        st.push(i);
    }

    int ans = 0;
    for (int i = 0; i < n; i++)
        ans = Math.max(ans, left[i] + right[i] + 1);

    return ans;
}